### PR TITLE
Sync stuff

### PR DESCRIPTION
**Pull request by ErikBjare**
Resuming work from #71 

What's missing:

 - [x] "2-way" test with actual folder 
 - [ ] Test for when events have been modified/deleted (when should things get synced?) 
 - [x] Use device IDs to identify remotes
    - Partially done, but can probably be done better.
 - [ ] refactor (some things can be broken down/reused)
 - [ ] better error handling (proper aw-client-rust errors, get rid of unwraps) 
 - [ ] more comprehensive tests (should be >90% covered)  
 - [ ] review processes/conventions
 - [ ] See comments in #71 and fix them
 - [ ] anything else?

### CODE REVIEW

**Comment by ErikBjare** [comment]
Rebasing on master somehow merged the PR. I assume the commit had been cherrypicked or something?

**Comment by ErikBjare** [comment]
Got sync working using aw-client-rust. I've only tested syncing temporary databases to my local testing instance, but it seems to work!

**Comment by ErikBjare** [comment]
For some reason can't request a review from @xylix, but would be nice to have as many eyes on this as possible.

**Comment by xylix** [comment]
Interesting that requesting doesn't work. I'll take a look around

**Comment by xylix** [comment]
The code looks pretty good. What are the currently missing features? Is "See comments in #71 and fix them" not complete yet?

**Comment by ErikBjare** [comment]
What's missing:

 - [ ] "2-way" test with actual folder 
 - [ ] better error handling (proper aw-client-rust errors, get rid of unwraps) 
 - [ ] more comprehensive tests (should be >90% covered) 
 - [ ] review processes/conventions
 - [ ] anything else?

**Comment by ErikBjare** [comment]
@johan-bjareholt I've stumbled into an issue where I need to know the device ID for a remote bucket... Could you take a look at it so we can discuss options?

File: aw-server/src/config.rs [hunk 1]
```diff
@@ -116,3 +118,21 @@ pub fn create_config(testing: bool) -> AWConfig {
 
     aw_config
 }
+
+/// Retrieves the device ID, if none exists it generates one (using UUID v4)
+pub fn get_device_id() -> String {
+    // TODO: Use seperate device_id for testing mode
+    // TODO: Cache to avoid retrieving on every /info call
```
Comment (johan-bjareholt) [inline aw-server/src/config.rs:125]: Why not handle it the same way as the config, read it once and then pass it around to those who need it? Makes more sense from a Rust standpoint and is less error prone.
Could possibly even be a part of the config structure if you wanted.

File: aw-server/src/config.rs [hunk 2]
```diff
@@ -116,3 +118,21 @@ pub fn create_config(testing: bool) -> AWConfig {
...
+    // TODO: How should these unwraps be removed?
```
Comment (johan-bjareholt) [inline aw-server/src/config.rs:126]: It should only be read once, and that's before the server starts.
Then it should simply be passed around like the config.

File: aw-sync/src/sync.rs [hunk 3]
```diff
@@ -85,134 +109,258 @@ impl AccessMethod for AwClient {
...
 }
 
-/// Performs a single sync pass
+/// Performs a single sync pass, in one direction
+///
+/// Steps:
+///   - Check the remotes
+///   - Create any new buckets (avoiding bucket ID conflicts by appending '-synced-from-{deviceid}')
+///   - Importing any new events
+#[allow(dead_code)]
```
Comment (johan-bjareholt) [inline aw-sync/src/sync.rs:145]: Why allow this? Why not just comment the dead code out?

Comment (ErikBjare) [reply aw-sync/src/sync.rs:145]: It's not dead code, I don't know why it is reported as such.

**Comment by ErikBjare** [comment]
@johan-bjareholt I think it's about time to merge this. There's work left but should be mergeable (assuming CI passes). Review plz.

File: aw-server/Cargo.toml [hunk 4]
```diff
@@ -28,7 +28,7 @@ fern = { version = "0.6", features = ["colored"] }
 toml = "0.5"
 gethostname = "0.2"
 uuid = { version = "0.8", features = ["serde", "v4"] }
-getopts = "0.2"
+clap = "3.0.0-beta.2"
```
Comment (johan-bjareholt) [inline aw-server/Cargo.toml:31]: I chose getops over clap due to getopts being so much leaner, clap has a silly amount of dependencies for just being a options parser.
Personally I don't think all those extra dependencies are worth the very small improvements that clap gives for our use-case.

File: aw-client-rust/src/lib.rs [hunk 5]
```diff
@@ -127,8 +128,8 @@ impl AwClient {
 
     pub fn get_event_count(&self, bucketname: &str) -> Result<i64, reqwest::Error> {
         let url = format!("{}/api/0/buckets/{}/events/count", self.baseurl, bucketname);
-        let res = self.client.get(&url).send()?.text()?;
-        let count: i64 = match res.parse() {
+        let res = self.client.get(&url).send()?.error_for_status()?.text()?;
+        let count: i64 = match res.trim().parse() {
```
[the remark this answers is not in the document]
Comment (ErikBjare) [reply aw-client-rust/src/lib.rs:132]: I think there was an extra newline that broke for either the Python or Rust server (probably the Python server).

File: aw-server/Cargo.toml (hunk 4 above)
Comment (ErikBjare) [reply aw-server/Cargo.toml:31]: Oh, didn't know that. 

I just find working with clap/clicky-like CLI builders a lot easier, the resulting CLI is better, and it makes the code simpler (but at the cost of dependencies).

**Comment by ErikBjare** [comment]
Merging this for now, as the CLI additions are needed for my PR in https://github.com/ErikBjare/quantifiedme/pull/4

Will later continue the work in a new PR.

**Pull request merged**